include/deal.II/fe/fe_enriched.h

```diff
@@ -691,6 +696,311 @@ class FE_Enriched : public FiniteElement<dim, spacedim>
+   * </tr>
+   * <tr>
+   * <td>1</td>
+   * <td><code>FE_Enriched<2>[FE_Q<2>(2)-FE_Q<2>(1)-FE_Nothing<2>(dominating)]</code></td>
+   * </tr>
+   * <tr>
+   * <td>2</td>
+   * <td><code>FE_Enriched<2>[FE_Q<2>(2)-FE_Q<2>(1)-FE_Q<2>(1)]</code></td>
+   * </tr>
+   * <tr>
+   * <td>3</td>
+   * <td><code>FE_Enriched<2>[FE_Q<2>(2)-FE_Nothing<2>(dominating)-FE_Q<2>(1)]</code></td>
+   * </tr>
+   * </table>
+   *
+   * The internal data members used by this class need to be available when the
+   * problem is solved. This can be ensured by declaring the object static,
+   * which is deallocated only when the program terminates. An alternative
+   * would be to use it as a data member of the containing class. Since vector
+   * of predicates and enrichment functions may not be available while
+   * constructing the Helper, a shared_ptr to Helper object can be used and
+   * constructed when the predicates and enrichment functions are
+   * available.
+   *
+   * <h3>Example usage:</h3>
+   * @code
+   * FE_Q<dim> fe_base(2);
+   * FE_Q<dim> fe_enriched(1);
+   * std::vector< predicate_function<dim> > predicates;
+   * std::vector< std::shared_ptr<Function<dim>> > enrichments;
+   * std::shared_ptr< hp::FECollection<dim> > fe_collection;
+   *
+   * Triangulation<dim>  triangulation;
+   * hp::DoFHandler<dim> dof_handler(triangulation);
+   *
+   * static ColorEnriched::Helper<dim> FE_helper(fe_base,
+   *                                             fe_enriched,
+   *                                             predicates,
+   *                                             enrichments);
+   * fe_collection = FE_helper.get_fe_collection(dof_handler);
+   * @endcode
+   */
+  template <int dim, int spacedim = dim>
+  struct Helper
+  {
+    /**
+     * Constructor for Helper class.
+     *
+     * @param fe_base A base FiniteElement
+     * @param fe_enriched An enriched FiniteElement
+     * @param predicates std::vector of predicates defining the sub-domains.
+     * <code>@p predicates[i]</code> returns true for a cell if it
+     * belongs to a sub-domain with index (i).
+     * @param enrichments std::vector of enrichment functions
+     */
+    Helper(const FE_Q<dim, spacedim> &                             fe_base,
+           const FE_Q<dim, spacedim> &                             fe_enriched,
+           const std::vector<predicate_function<dim, spacedim>> &  predicates,
+           const std::vector<std::shared_ptr<Function<spacedim>>> &enrichments);
+
+    /**
+     * Prepares an hp::DoFHandler object. The active FE indices of
+     * mesh cells are initialized to work with
+     * ColorEnriched::Helper<dim,spacedim>::fe_collection.
+     *
+     * Returns an hp::FECollection object.
+     *
+     * @param dof_handler an hp::DoFHandler object
+     * @return shared pointer to hp::FECollection, a collection of
+     * finite elements needed by @p dof_handler.
+     */
+    std::shared_ptr<hp::FECollection<dim, spacedim>>
```

Comment:
i thought you wanted to return a `const hp::FECollection<dim, spacedim>&` ?

Reply:
i wanted to but FECollection only has 
operator= (FECollection< dim, spacedim > &&)=default.

So i just used shared pointer.

Reply:
Then you probably forgot to add the ampersand at the end. You may not be able to return it as a value, but you should always be able to return it as a `const` reference.

Reply:
ok

Reply:
i don't have the commits, but I think I used const reference &. The function return was not an issue but when I wrote 
(FECollection) = (function return) 
It says it couldn't find the function operator= for const reference & argument!

I will check again and get back to you.

Reply:
Then, you are invoking the copy constructor somewhere.

Reply:
Yes -- you need to use the result to initialize a reference variable, not later to copy by value. So this will work:
```
  const FECollection &fe_collection = object.get_fe_collection(...);
```
because it initializes a reference variable. But this will fails:
```
  FECollection fe_collection;
  fe_collection = object.get_fe_collection(...);
```
because it copies a value, which you can't do here because `FECollection` does not have a copy operator.

Reply:
Now i see what you meant. I will try changing it.